### ISSUE TITLE
[Discussion] Implementing Wiener Channel Filter in TensorFlow

### ISSUE THREAD

**Issue by RomanScott** [deezer/spleeter#127]
Hello,

As part of the separation process, Spleeter applies a multichannel Wiener Filter via the Norbert library. However, Norbert runs off the CPU, thus making it relatively slower compared to a GPU.

Is there a way to port the Norbert algorithm to TensorFlow? I know there is a [PyTorch implementation of Norbert](https://github.com/sigsep/open-unmix-pytorch/blob/torchfilters/filtering.py) in the [torchfilters branch](https://github.com/sigsep/open-unmix-pytorch/tree/torchfilters) of Open-Unmix.

It seems to rely on an inverse STFT, which TensorFlow implements. Hopefully this can be achieved, but if not that's okay.

Tagging @faroit because I know he was working on a TensorFlow version of Open-Unmix a while back.

Thank you!

**Comment by mmoussallam** [comment]
Hi Roman,

Probably you want to ask that to the good folks maintaining norbert. The MWF is actually a more complex than just applying mask + STFT and typically requires an EM algorithm to refine the initial mask estimates. I'm not a specialist here so it's not obvious to me if a GPU porting of this EM phase would be faster.

**Comment by RomanScott** [comment]
@mmoussallam 

Thank you. I am tagging @aliutkus because I think he wrote the Norbert library. He might have an unrealeased Tensorflow version?

**Comment by faroit** [comment]
we decided to not release the tensorflow implementation of open-unmix publicly, but rather to make it available for commercial applications. We can therefore not share the tf version of norbert.

@RomanScott Feel free to convert the torch code to tensorflow. This is def possible since tf provides almost the same sets of operators needed for this functions. In case of spleeter which runs a joint training where the mask is softmax activated the MWF might only improve the results a little as stated in the spleeter documentation.

**Comment by aliutkus** [comment]
@RomanScott I should add that I did the port of `norbert` to tensorflow for my private work, and it goes smoothly, especially considering that tensorflow does support complex tensors

good luck

**Comment by mmoussallam** [comment]
Closing this issue as there's not much more we can do here.

**Comment by Danngonz3** [comment]
@RomanScott would you be willing to share your tf implementation of the wiener filter? 

I’ve looked into doing it myself, however since tf doesn’t support slice assignment, it is a bit tricky.
